README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Asynchronous Messaging for Pip.Services in Node.js / ES2017ß

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The Messaging module contains a set of interfaces and classes for working with message queues, as well as an in-memory message queue implementation. 

The module contains the following packages:

- **Build** - in-memory message queue factory
- **Queues** - contains interfaces for working with message queues, subscriptions for receiving messages from the queue, and an in-memory message queue implementation.

<a name="links"></a> Quick links:

* [Configuration](http://docs.pipservices.org/toolkit/getting_started/configurations/)
* [API Reference](https://pip-services3-nodex.github.io/pip-services3-messaging-nodex/globals.html)
* [Change Log](CHANGELOG.md)
* [Get Help](http://docs.pipservices.org/get_help/)
* [Contribute](http://docs.pipservices.org/contribute/)

## Use

Install the NPM package as
```bash
npm install pip-services3-messaging-nodex --save
```

## Develop

For development you shall install the following prerequisites:
* Node.js 8+
* Visual Studio Code or another IDE of your choice
* Docker
* Typescript

Install dependencies:
```bash
npm install
```

Compile the code:
```bash
tsc
```

Run automated tests:
```bash
npm test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The Node.js version of Pip.Services is created and maintained by:
- **Volodymyr Tkachenko**
- **Sergey Seroukhov**

The documentation is written by:
- **Mark Makarychev**

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Asynchronous Messaging for Pip.Services in Node.js / ES2017 Changelog

## <a name="1.0.1"></a> 1.0.1 (2023-02-07)

### Features
* Updated dependencies

## <a name="1.0.0"></a> 1.0.0 (2021-04-09)

Initial public release



src/test/TestMessageReceiver.ts
/** @module test */

import { ICleanable } from 'pip-services3-commons-nodex';

import { IMessageReceiver } from '../queues/IMessageReceiver';
import { IMessageQueue } from '../queues/IMessageQueue';
import { MessageEnvelope } from '../queues/MessageEnvelope';

export class TestMessageReceiver implements IMessageReceiver, ICleanable {
    private _messages: MessageEnvelope[] = [];

    constructor() {}

    /**
     * Gets the list of received messages.
     */
    public get messages(): MessageEnvelope[] {
        return this._messages;
    }

    /**
     * Gets the received message count.
     */
    public get messageCount(): number {
        return this._messages.length;
    }

    /**
     * Receives incoming message from the queue.
     * 
     * @param envelope  an incoming message
     * @param queue     a queue where the message comes from
     * 
     * @see [[MessageEnvelope]]
     * @see [[IMessageQueue]]
     */
    public async receiveMessage(envelope: MessageEnvelope, queue: IMessageQueue): Promise<void> {
        this._messages.push(envelope);
    }

    /**
     * Clears all received messagers.
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     */
    public async clear(correlationId: string): Promise<void> {
        this._messages = [];
    }
}

test/queues/MessageQueueFixture.ts
const assert = require('chai').assert;

import { IMessageQueue } from '../../src/queues/IMessageQueue';
import { MessageEnvelope } from '../../src/queues/MessageEnvelope';
import { RandomString, IdGenerator } from 'pip-services3-commons-nodex';
import { TestMessageReceiver } from '../../src/test/TestMessageReceiver';

export class MessageQueueFixture {
    private _queue: IMessageQueue;

    public constructor(queue: IMessageQueue) {
        this._queue = queue;
    }

    public async testSendReceiveMessage(): Promise<void> {
        let envelope1: MessageEnvelope = new MessageEnvelope("123", "Test", "Test message");
        await this._queue.send(null, envelope1);

        let envelope2 = await this._queue.receive(null, 10000);
        assert.isNotNull(envelope2);
        assert.equal(envelope1.message_type, envelope2.message_type);
        assert.equal(envelope1.message, envelope2.message);
        assert.equal(envelope1.correlation_id, envelope2.correlation_id);
    }

    public async testReceiveSendMessage() {
        let envelope1: MessageEnvelope = new MessageEnvelope("123", "Test", "Test message");

        setTimeout(() => {
            this._queue.send(null, envelope1);
        }, 500);

        let envelope2 = await this._queue.receive(null, 10000);
        assert.isNotNull(envelope2);
        assert.equal(envelope1.message_type, envelope2.message_type);
        assert.equal(envelope1.message, envelope2.message);
        assert.equal(envelope1.correlation_id, envelope2.correlation_id);
    }

    public async testReceiveCompleteMessage() {
        let envelope1: MessageEnvelope = new MessageEnvelope("123", "Test", "Test message");

        await this._queue.send(null, envelope1);

        let count = await this._queue.readMessageCount();
        assert.isTrue(count > 0);

        let envelope2 = await this._queue.receive(null, 10000);
        assert.isNotNull(envelope2);
        assert.equal(envelope1.message_type, envelope2.message_type);
        assert.equal(envelope1.message, envelope2.message);
        assert.equal(envelope1.correlation_id, envelope2.correlation_id);

        await this._queue.complete(envelope2);
        assert.isNull(envelope2.getReference());
    }

    public async testReceiveAbandonMessage() {
        let envelope1: MessageEnvelope = new MessageEnvelope("123", "Test", "Test message");
        await this._queue.send(null, envelope1);

        let envelope2 = await this._queue.receive(null, 10000);
        assert.isNotNull(envelope2);
        assert.equal(envelope1.message_type, envelope2.message_type);
        assert.equal(envelope1.message, envelope2.message);
        assert.equal(envelope1.correlation_id, envelope2.correlation_id);

        await this._queue.abandon(envelope2);

        envelope2 = await this._queue.receive(null, 10000);
        assert.isNotNull(envelope2);
        assert.equal(envelope1.message_type, envelope2.message_type);
        assert.equal(envelope1.message, envelope2.message);
        assert.equal(envelope1.correlation_id, envelope2.correlation_id);
    }

    public async testSendPeekMessage() {
        let envelope1: MessageEnvelope = new MessageEnvelope("123", "Test", "Test message");
        await this._queue.send(null, envelope1);

        let envelope2 = await this._queue.peek(null);
        assert.isNotNull(envelope2);
        assert.equal(envelope1.message_type, envelope2.message_type);
        assert.equal(envelope1.message, envelope2.message);
        assert.equal(envelope1.correlation_id, envelope2.correlation_id);
    }

    public async testPeekNoMessage() {
        let envelope = await this._queue.peek(null);
        assert.isNull(envelope);
    }

    public async testMoveToDeadMessage() {
        let envelope1: MessageEnvelope = new MessageEnvelope("123", "Test", "Test message");
        await this._queue.send(null, envelope1);

        let envelope2 = await this._queue.receive(null, 10000);
        assert.isNotNull(envelope2);
        assert.equal(envelope1.message_type, envelope2.message_type);
        assert.equal(envelope1.message, envelope2.message);
        assert.equal(envelope1.correlation_id, envelope2.correlation_id);

        await this._queue.moveToDeadLetter(envelope2);
    }

    public async testOnMessage() {
        let messageReceiver = new TestMessageReceiver();
        this._queue.beginListen(null, messageReceiver);

        await new Promise<void>((resolve, reject) => setTimeout(resolve, 1000));

        let envelope1: MessageEnvelope = new MessageEnvelope("123", "Test", "Test message");
        await this._queue.send(null, envelope1);

        await new Promise<void>((resolve, reject) => setTimeout(resolve, 1000));

        let envelope2 = messageReceiver.messages[0];
        assert.isNotNull(envelope2);
        assert.equal(envelope1.message_type, envelope2.message_type);
        assert.equal(envelope1.message, envelope2.message);
        assert.equal(envelope1.correlation_id, envelope2.correlation_id);

        this._queue.endListen(null);
    }

    public async testSendAsObject() {
        let messageReceiver = new TestMessageReceiver();
        let testObj = {
            id: IdGenerator.nextLong(),
            name: RandomString.nextString(20, 50)
        };

        this._queue.beginListen(null, messageReceiver);

        await new Promise<void>((resolve, reject) => setTimeout(resolve, 1000));

        //  send array of strings 
        await this._queue.sendAsObject('123', 'messagetype', ['string1', 'string2']);

        await new Promise<void>((resolve, reject) => setTimeout(resolve, 1000));

        assert.equal(1, messageReceiver.messageCount);
        let envelope = messageReceiver.messages[0];
        assert.isNotNull(envelope);
        assert.equal('messagetype', envelope.message_type);
        assert.equal('123', envelope.correlation_id);

        let message = envelope.getMessageAs<string[]>();
        assert.isArray(message);
        assert.includeMembers(message, ['string1', 'string2']);

        // send string
        await messageReceiver.clear(null);
        await this._queue.sendAsObject('123', 'messagetype', 'string2');

        await new Promise<void>((resolve, reject) => setTimeout(resolve, 1000));

        assert.equal(1, messageReceiver.messageCount);
        envelope = messageReceiver.messages[0];
        assert.isNotNull(envelope);
        assert.equal('messagetype', envelope.message_type);
        assert.equal('123', envelope.correlation_id);

        let message2 = envelope.getMessageAsString();
        assert.equal('string2', message2);

        // // send object
        // await messageReceiver.clear(null);
        // await this._queue.sendAsObject('123', 'messagetype', testObj);

        // await new Promise<void>((resolve, reject) => setTimeout(resolve, 1000));

        // assert.equal(1, messageReceiver.messageCount);
        // envelope = messageReceiver.messages[0];
        // assert.isNotNull(envelope);
        // assert.equal('messagetype', envelope.message_type);
        // assert.equal('123', envelope.correlation_id);

        // let message3 = envelope.getMessageAs<any>();
        // assert.isNotNull(message);
        // assert.equal(testObj.id, message3.id);
        // assert.equal(testObj.name, message3.name);

        this._queue.endListen(null);
    }

}


test/queues/MemoryMessageQueue.test.ts
import { MemoryMessageQueue } from '../../src/queues/MemoryMessageQueue';
import { MessageQueueFixture } from './MessageQueueFixture';

suite('MemoryMessageQueue', () => {
    let queue: MemoryMessageQueue;
    let fixture: MessageQueueFixture;

    suiteSetup(async () => {
        queue = new MemoryMessageQueue("TestQueue");
        fixture = new MessageQueueFixture(queue);
        await queue.open(null);
    });

    suiteTeardown(async () => {
        await queue.close(null);
    });

    setup(async () => {
        await queue.clear(null);
    });

    test('Send Receive Message', async () => {
        await fixture.testSendReceiveMessage();
    });

    test('Receive Send Message', async () => {
        await fixture.testReceiveSendMessage();
    });

    test('Receive And Complete Message', async () => {
        await fixture.testReceiveCompleteMessage();
    });

    test('Receive And Abandon Message', async () => {
        await fixture.testReceiveAbandonMessage();
    });

    test('Send Peek Message', async () => {
        await fixture.testSendPeekMessage();
    });

    test('Peek No Message', async () => {
        await fixture.testPeekNoMessage();
    });

    test('Move To Dead Message', async () => {
        await fixture.testMoveToDeadMessage();
    });

    test('On Message', async () => {
        await fixture.testOnMessage();
    });

    test('Send Message As Object', async () => {
        await fixture.testSendAsObject();
    });
});


src/queues/IMessageQueue.ts
/** @module queues */
import { IOpenable } from 'pip-services3-commons-nodex';
import { IClosable } from 'pip-services3-commons-nodex';

import { MessagingCapabilities } from './MessagingCapabilities';
import { MessageEnvelope } from './MessageEnvelope';
import { IMessageReceiver } from './IMessageReceiver';

/**
 * Interface for asynchronous message queues.
 * 
 * Not all queues may implement all the methods.
 * Attempt to call non-supported method will result in NotImplemented exception.
 * To verify if specific method is supported consult with [[MessagingCapabilities]].
 * 
 * @see [[MessageEnvelope]]
 * @see [[MessagingCapabilities]]
 */
export interface IMessageQueue extends IOpenable, IClosable {
    /**
     * Gets the queue name
     * 
     * @returns the queue name.
     */
    getName(): string;

    /**
     * Gets the queue capabilities
     * 
     * @returns the queue's capabilities object.
     */
	getCapabilities(): MessagingCapabilities;
    
    /**
     * Reads the current number of messages in the queue to be delivered.
     * 
     * @returns      a number of messages.
     */
    readMessageCount(): Promise<number>;

    /**
     * Sends a message into the queue.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param envelope          a message envelop to be sent.
     */
    send(correlationId: string, envelope: MessageEnvelope): Promise<void>;

    /**
     * Sends an object into the queue.
     * Before sending the object is converted into JSON string and wrapped in a [[MessageEnvelope]].
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param messageType       a message type
     * @param value             an object value to be sent
     * 
     * @see [[send]]
     */
    sendAsObject(correlationId: string, messageType: string, value: any): Promise<void>;

    /**
     * Peeks a single incoming message from the queue without removing it.
     * If there are no messages available in the queue it returns null.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @returns                 a peeked message or <code>null</code>.
     */
    peek(correlationId: string): Promise<MessageEnvelope>;

    /**
     * Peeks multiple incoming messages from the queue without removing them.
     * If there are no messages available in the queue it returns an empty list.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param messageCount      a maximum number of messages to peek.
     * @returns                 a peeked list with messages.
     */
    peekBatch(correlationId: string, messageCount: number): Promise<MessageEnvelope[]>

    /**
     * Receives an incoming message and removes it from the queue.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param waitTimeout       a timeout in milliseconds to wait for a message to come.
     * @returns                 a received message or <code>null</code>.
     */
    receive(correlationId: string, waitTimeout: number): Promise<MessageEnvelope>;

    /**
     * Renews a lock on a message that makes it invisible from other receivers in the queue.
     * This method is usually used to extend the message processing time.
     * 
     * @param message       a message to extend its lock.
     * @param lockTimeout   a locking timeout in milliseconds.
     */
    renewLock(message: MessageEnvelope, lockTimeout: number): Promise<void>;

    /**
     * Permanently removes a message from the queue.
     * This method is usually used to remove the message after successful processing.
     * 
     * @param message   a message to remove.
     */
    complete(message: MessageEnvelope): Promise<void>;

    /**
     * Returnes message into the queue and makes it available for all subscribers to receive it again.
     * This method is usually used to return a message which could not be processed at the moment
     * to repeat the attempt. Messages that cause unrecoverable errors shall be removed permanently
     * or/and send to dead letter queue.
     * 
     * @param message   a message to return.
     */
    abandon(message: MessageEnvelope): Promise<void>; 

    /**
     * Permanently removes a message from the queue and sends it to dead letter queue.
     * 
     * @param message   a message to be removed.
     */
    moveToDeadLetter(message: MessageEnvelope): Promise<void>;

    /**
     * Listens for incoming messages and blocks the current thread until queue is closed.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param receiver          a receiver to receive incoming messages.
     * 
     * @see [[IMessageReceiver]]
     * @see [[receive]]
     */
    listen(correlationId: string, receiver: IMessageReceiver): void;

    /**
     * Listens for incoming messages without blocking the current thread.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param receiver          a receiver to receive incoming messages.
     * 
     * @see [[listen]]
     * @see [[IMessageReceiver]]
     */
    beginListen(correlationId: string, receiver: IMessageReceiver): void;

    /**
     * Ends listening for incoming messages.
     * When this method is call [[listen]] unblocks the thread and execution continues.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     */
    endListen(correlationId: string): void;
}

src/queues/IMessageReceiver.ts
/** @module queues */
import { IMessageQueue } from './IMessageQueue';
import { MessageEnvelope } from './MessageEnvelope';

/**
 * Callback interface to receive incoming messages.
 * 
 * ### Example ###
 * 
 *     class MyMessageReceiver implements IMessageReceiver {
 *       public async receiveMessage(envelop: MessageEnvelop, queue: IMessageQueue): Promise<void> {
 *           console.log("Received message: " + envelop.getMessageAsString());
 *       }
 *     }
 * 
 *     let messageQueue = new MemoryMessageQueue();
 *     messageQueue.listen("123", new MyMessageReceiver());
 * 
 *     await messageQueue.open("123")
 *     await messageQueue.send("123", new MessageEnvelop(null, "mymessage", "ABC")); // Output in console: "ABC"
 */
export interface IMessageReceiver {
    /**
     * Receives incoming message from the queue.
     * 
     * @param envelope  an incoming message
     * @param queue     a queue where the message comes from
     * 
     * @see [[MessageEnvelope]]
     * @see [[IMessageQueue]]
     */
    receiveMessage(envelope: MessageEnvelope, queue: IMessageQueue): Promise<void>;
}

src/queues/MemoryMessageQueue.ts
/** @module queues */
import { ConnectionParams } from 'pip-services3-components-nodex';
import { CredentialParams } from 'pip-services3-components-nodex';
import { ConfigParams } from 'pip-services3-commons-nodex';

import { IMessageReceiver } from './IMessageReceiver';
import { MessageQueue } from './MessageQueue';
import { MessageEnvelope } from './MessageEnvelope';
import { MessagingCapabilities } from './MessagingCapabilities';
import { LockedMessage } from './LockedMessage';

/**
 * Message queue that sends and receives messages within the same process by using shared memory.
 * 
 * This queue is typically used for testing to mock real queues.
 * 
 * ### Configuration parameters ###
 * 
 * - name:                        name of the message queue
 * 
 * ### References ###
 * 
 * - <code>\*:logger:\*:\*:1.0</code>           (optional) [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/log.ilogger.html ILogger]] components to pass log messages
 * - <code>\*:counters:\*:\*:1.0</code>         (optional) [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/count.icounters.html ICounters]] components to pass collected measurements
 * 
 * @see [[MessageQueue]]
 * @see [[MessagingCapabilities]]
 * 
 * ### Example ###
 * 
 *     let queue = new MessageQueue("myqueue");
 * 
 *     await queue.send("123", new MessageEnvelop(null, "mymessage", "ABC"));
 * 
 *     let message = await queue.receive("123");
 *     if (message != null) {
 *        ...
 *        await queue.complete("123", message);
 *     }
 */
export class MemoryMessageQueue extends MessageQueue {
    private _messages: MessageEnvelope[] = [];
    private _lockTokenSequence: number = 0;
    private _lockedMessages: { [id: number]: LockedMessage; } = {};
    private _opened: boolean = false;
    /** Used to stop the listening process. */
    private _cancel: boolean = false;
    private _listenInterval: number = 1000;

    /**
     * Creates a new instance of the message queue.
     * 
     * @param name  (optional) a queue name.
     * 
     * @see [[MessagingCapabilities]]
     */
    public constructor(name?: string) {
        super(name);
        this._capabilities = new MessagingCapabilities(true, true, true, true, true, true, true, false, true);
    }

    /**
	 * Checks if the component is opened.
	 * 
	 * @returns true if the component has been opened and false otherwise.
     */
    public isOpen(): boolean {
        return this._opened;
    }

    /**
     * Opens the component with given connection and credential parameters.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param connections        connection parameters
     * @param credential        credential parameters
     */
    protected async openWithParams(correlationId: string, connections: ConnectionParams[],
        credential: CredentialParams): Promise<void> {
        this._opened = true;
    }

    /**
	 * Closes component and frees used resources.
	 * 
	 * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async close(correlationId: string): Promise<void> {
        this._opened = false;
        this._cancel = true;
        this._logger.trace(correlationId, "Closed queue %s", this);
    }

    /**
	 * Clears component state.
	 * 
	 * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async clear(correlationId: string): Promise<void> {
        this._messages = [];
        this._lockedMessages = {};
        this._cancel = false;
    }
    
    /**
     * Configures component by passing configuration parameters.
     * 
     * @param config    configuration parameters to be set.
     */
    public configure(config: ConfigParams): void {
        super.configure(config);

        this._listenInterval = config.getAsIntegerWithDefault('listen_interval', this._listenInterval);
        this._listenInterval = config.getAsIntegerWithDefault('options.listen_interval', this._listenInterval);
    }

    /**
     * Reads the current number of messages in the queue to be delivered.
     * 
     * @returns     a number of messages in the queue.
     */
    public async readMessageCount(): Promise<number> {
        return this._messages.length;
    }

    /**
     * Sends a message into the queue.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param envelope          a message envelop to be sent.
     */
    public async send(correlationId: string, envelope: MessageEnvelope): Promise<void> {
        envelope.sent_time = new Date();

        // Add message to the queue
        this._messages.push(envelope);

        this._counters.incrementOne("queue." + this.getName() + ".sent_messages");
        this._logger.debug(envelope.correlation_id, "Sent message %s via %s", envelope.toString(), this.getName());
    }

    /**
     * Peeks a single incoming message from the queue without removing it.
     * If there are no messages available in the queue it returns null.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @returns                 a peeked message or <code>null</code>.
     */
    public async peek(correlationId: string): Promise<MessageEnvelope> {
        let message: MessageEnvelope = null;

        // Pick a message
        if (this._messages.length > 0)
            message = this._messages[0];

        if (message != null)
            this._logger.trace(message.correlation_id, "Peeked message %s on %s", message, this.getName());

        return message;
    }

    /**
     * Peeks multiple incoming messages from the queue without removing them.
     * If there are no messages available in the queue it returns an empty list.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param messageCount      a maximum number of messages to peek.
     * @returns                 a list with peeked messages.
     */
    public async peekBatch(correlationId: string, messageCount: number): Promise<MessageEnvelope[]>{
        let messages = this._messages.slice(0, messageCount);
        
        this._logger.trace(correlationId, "Peeked %d messages on %s", messages.length, this.getName());
    
        return messages;
    }

    /**
     * Receives an incoming message and removes it from the queue.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param waitTimeout       a timeout in milliseconds to wait for a message to come.
     * @returns                 a received message or <code>null</code>.
     */
    public async receive(correlationId: string, waitTimeout: number): Promise<MessageEnvelope> {
        let checkIntervalMs = 100;
        let elapsedTime = 0;

        // Get message the the queue
        let message = this._messages.shift();

        while (elapsedTime < waitTimeout && message == null) {
            // Wait for a while
            await new Promise((resolve, reject) => { setTimeout(resolve, checkIntervalMs); });
            elapsedTime += checkIntervalMs;

            // Get message the the queue
            message = this._messages.shift();
        }

        if (message == null) {
            return message;
        }

        // Generate and set locked token
        let lockedToken = this._lockTokenSequence++;
        message.setReference(lockedToken);

        // Add messages to locked messages list
        let lockedMessage: LockedMessage = new LockedMessage();
        let now: Date = new Date();
        now.setMilliseconds(now.getMilliseconds() + waitTimeout);
        lockedMessage.expirationTime = now;
        lockedMessage.message = message;
        lockedMessage.timeout = waitTimeout;
        this._lockedMessages[lockedToken] = lockedMessage;

        // Instrument the process
        this._counters.incrementOne("queue." + this.getName() + ".received_messages");
        this._logger.debug(message.correlation_id, "Received message %s via %s", message, this.getName());

        return message;
    }

    /**
     * Renews a lock on a message that makes it invisible from other receivers in the queue.
     * This method is usually used to extend the message processing time.
     * 
     * @param message       a message to extend its lock.
     * @param lockTimeout   a locking timeout in milliseconds.
     */
    public async renewLock(message: MessageEnvelope, lockTimeout: number): Promise<void> {
        if (message.getReference() == null) {
            return;
        }

        // Get message from locked queue
        let lockedToken: number = message.getReference();
        let lockedMessage: LockedMessage = this._lockedMessages[lockedToken];

        // If lock is found, extend the lock
        if (lockedMessage) {
            let now: Date = new Date();
            // Todo: Shall we skip if the message already expired?
            if (lockedMessage.expirationTime > now) {
                now.setMilliseconds(now.getMilliseconds() + lockedMessage.timeout);
                lockedMessage.expirationTime = now;
            }
        }

        this._logger.trace(message.correlation_id, "Renewed lock for message %s at %s", message, this.getName());
    }

    /**
     * Permanently removes a message from the queue.
     * This method is usually used to remove the message after successful processing.
     * 
     * @param message   a message to remove.
     */
    public async complete(message: MessageEnvelope): Promise<void> {
        if (message.getReference() == null) {
            return;
        }

        let lockKey: number = message.getReference();
        delete this._lockedMessages[lockKey];
        message.setReference(null);

        this._logger.trace(message.correlation_id, "Completed message %s at %s", message, this.getName());
    }

    /**
     * Returnes message into the queue and makes it available for all subscribers to receive it again.
     * This method is usually used to return a message which could not be processed at the moment
     * to repeat the attempt. Messages that cause unrecoverable errors shall be removed permanently
     * or/and send to dead letter queue.
     * 
     * @param message   a message to return.
     */
    public async abandon(message: MessageEnvelope): Promise<void> {
        if (message.getReference() == null) {
            return;
        }

        // Get message from locked queue
        let lockedToken: number = message.getReference();
        let lockedMessage: LockedMessage = this._lockedMessages[lockedToken];
        if (lockedMessage) {
            // Remove from locked messages
            delete this._lockedMessages[lockedToken];
            message.setReference(null);

            // Skip if it is already expired
            if (lockedMessage.expirationTime <= new Date()) {
                return;
            }
        }
        // Skip if it absent
        else {
            return;
        }

        this._logger.trace(message.correlation_id, "Abandoned message %s at %s", message, this.getName());

        await this.send(message.correlation_id, message);
    }

    /**
     * Permanently removes a message from the queue and sends it to dead letter queue.
     * 
     * @param message   a message to be removed.
     */
    public async moveToDeadLetter(message: MessageEnvelope): Promise<void> {
        if (message.getReference() == null) {
            return;
        }

        let lockedToken: number = message.getReference();
        delete this._lockedMessages[lockedToken];
        message.setReference(null);

        this._counters.incrementOne("queue." + this.getName() + ".dead_messages");
        this._logger.trace(message.correlation_id, "Moved to dead message %s at %s", message, this.getName());
    }

    /**
     * Listens for incoming messages and blocks the current thread until queue is closed.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param receiver          a receiver to receive incoming messages.
     * 
     * @see [[IMessageReceiver]]
     * @see [[receive]]
     */
    public listen(correlationId: string, receiver: IMessageReceiver): void {
        let listenFunc = async () => {
            let timeoutInterval = this._listenInterval;

            this._logger.trace(null, "Started listening messages at %s", this.toString());

            this._cancel = false;

            while (!this._cancel) {
                try {
                    let message: MessageEnvelope = await this.receive(correlationId, timeoutInterval);
                    if (message != null && !this._cancel) {
                        await receiver.receiveMessage(message, this);
                    }
                } catch (ex) {
                    this._logger.error(correlationId, ex, "Failed to process the message");
                }
            }
        };        
        listenFunc();
    }

    /**
     * Ends listening for incoming messages.
     * When this method is call [[listen]] unblocks the thread and execution continues.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     */
    public endListen(correlationId: string): void {
        this._cancel = true;
    }

}